Exception not handled when ProcessWithResult.exec is executed with ProcessWithResult.runWithIO effect

I was trying out the code from https://doc.flix.dev/library-effects.html#processwithresult
and thought would test the error case with following code

```
def main(): Unit \ {Exec, IO} =
    run {
        match ProcessWithResult.exec("ls1", Nil) {
            case Result.Ok(_)    => println("Booyah")
            case Result.Err(err) => println("Unable to execute process: ${err}")
        }
    } with ProcessWithResult.runWithIO
```

It seems the error is not handled by case Result.Err(err) => .... as the REPL prints the following:

```
flix> main()
java.io.IOException: Cannot run program "ls1": error=2, No such file or directory
        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1110)
        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1044)
        at ProcessWithResult.Clo$handle$879937.applyFrame(Unknown Source)
        at ProcessWithResult.Clo$handle$879937.invoke(Unknown Source)
        at Eff$ProcessWithResult.execWithCwdAndEnv(Unknown Source)
        at dev.flix.runtime.Handler$.installHandler(dev/flix/runtime/Handler$)
        at ProcessWithResult.Clo$handle$879940.applyFrame(Unknown Source)
        at ProcessWithResult.Clo$handle$879940.invoke(Unknown Source)
        at Def$shell1.applyFrame(Unknown Source)
        at Def$shell1.invoke(Unknown Source)
        at Main.main(Main)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:565)
        at ca.uwaterloo.flix.language.phase.jvm.Loader$.mainFunction$1(Loader.scala:70)
        at ca.uwaterloo.flix.language.phase.jvm.Loader$.$anonfun$load$9(Loader.scala:79)
        at ca.uwaterloo.flix.language.phase.jvm.Loader$.$anonfun$load$9$adapted(Loader.scala:79)
        at ca.uwaterloo.flix.runtime.shell.Shell.run(Shell.scala:408)
        at ca.uwaterloo.flix.runtime.shell.Shell.execEval(Shell.scala:320)
        at ca.uwaterloo.flix.runtime.shell.Shell.execute(Shell.scala:166)
        at ca.uwaterloo.flix.runtime.shell.Shell.loop(Shell.scala:115)
        at ca.uwaterloo.flix.Main$.main(Main.scala:261)
        at ca.uwaterloo.flix.Main.main(Main.scala)
Caused by: java.io.IOException: error=2, No such file or directory
        at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
        at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:290)
        at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:221)
        at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1076)
        ... 21 more
```

Flix version : `0.58.1`
Editor : Vs code

--- a/main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala
+++ b/main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala
@@ -1276,9 +1276,6 @@ object GenExpression {
       // Introduce a label after the finally block.
       val afterFinally = new Label()
 
-      // Emit try finally block.
-      mv.visitTryCatchBlock(beforeTryBlock, afterTryBlock, finallyBlock, null)
-
       // Create an instance of Region
       mv.visitTypeInsn(NEW, BackendObjType.Region.jvmName.toInternalName)
       mv.visitInsn(DUP)
@@ -1291,6 +1288,10 @@ object GenExpression {
       // Compile the scope body
       mv.visitLabel(beforeTryBlock)
       compileExpr(exp)
+
+      // Emit try finally block. It's important to do this after compiling sub-expressions to ensure
+      // correct catch case ordering.
+      mv.visitTryCatchBlock(beforeTryBlock, afterTryBlock, finallyBlock, null)
 
       // When we exit the scope, call the region's `exit` method
       val iLoad = AsmOps.getLoadInstruction(JvmType.Reference(BackendObjType.Region.jvmName))
@@ -1334,11 +1335,6 @@ object GenExpression {
         rule => rule -> new Label()
       }
 
-      // Emit a try catch block for each catch rule.
-      for ((CatchRule(_, clazz, _), handlerLabel) <- rulesAndLabels) {
-        mv.visitTryCatchBlock(beforeTryBlock, afterTryBlock, handlerLabel, asm.Type.getInternalName(clazz))
-      }
-
       // Emit code for the try block.
       mv.visitLabel(beforeTryBlock)
       compileExpr(exp)
@@ -1357,6 +1353,12 @@ object GenExpression {
         // Emit code for the handler body expression.
         compileExpr(body)
         mv.visitJumpInsn(GOTO, afterTryAndCatch)
+      }
+
+      // Emit a try catch block for each catch rule. It's important to do this after compiling
+      // sub-expressions to ensure correct catch case ordering.
+      for ((CatchRule(_, clazz, _), handlerLabel) <- rulesAndLabels) {
+        mv.visitTryCatchBlock(beforeTryBlock, afterTryBlock, handlerLabel, asm.Type.getInternalName(clazz))
       }
 
       // Add the label after both the try and catch rules.

--- a/main/test/flix/Test.Exp.Regions.flix
+++ b/main/test/flix/Test.Exp.Regions.flix
@@ -1,4 +1,5 @@
 mod Test.Exp.Regions {
+    import java.lang.NullPointerException
 
     // Uses references and regions (e.g. could just duplicate most test cases for both).
     // Test cases that use 0, 1, 2, and 3 nested regions.

--- a/main/test/flix/Test.Exp.Regions.flix
+++ b/main/test/flix/Test.Exp.Regions.flix
@@ -1383,4 +1383,18 @@ mod Test.Exp.Regions {
             }
         }
 
+    @test
+    def testRegionsAndTryCatch(): Bool \ IO = {
+        region rc {
+            let res = Ref.fresh(rc, false);
+            try {
+                throw new NullPointerException("test");
+                Ref.put(false, res)
+            } catch {
+                case _: NullPointerException => Ref.put(true, res)
+            };
+            Ref.get(res)
+        }
+    }
+
 }

